pkg/apis/core/validation/validation.go

```diff
@@ -4668,6 +4674,37 @@ func ValidatePodUpdate(newPod, oldPod *core.Pod, opts PodValidationOptions) fiel
 		mungedPodSpec.TerminationGracePeriodSeconds = oldPod.Spec.TerminationGracePeriodSeconds // +k8s:verify-mutation:reason=clone
 	}
 
+	podIsGated := len(oldPod.Spec.SchedulingGates) > 0
+	if opts.AllowMutableNodeSelectorAndNodeAffinity && podIsGated {
+		// Additions to spec.nodeSelector are allowed (no deletions or mutations) for gated pods.
+		if !apiequality.Semantic.DeepEqual(oldPod.Spec.NodeSelector, mungedPodSpec.NodeSelector) {
+			allErrs = append(allErrs, validateNodeSelectorMutation(specPath.Child("nodeSelector"), oldPod.Spec.NodeSelector, mungedPodSpec.NodeSelector)...)
+			mungedPodSpec.NodeSelector = oldPod.Spec.NodeSelector // +k8s:verify-mutation:reason=clone
+		}
+
+		if oldPod.Spec.Affinity != nil && mungedPodSpec.Affinity != nil {
```

Comment:
it's hard to follow the munging/validation being conditional on whether the new and old top-level affinity fields are nil, when those _also_ contain pod affinity/anti-affinity which are not allowed to mutate

A structure like this might be easier to follow:

```go
var oldNodeAffinity *core.NodeAffinity
if oldPod.Spec.Affinity != nil {
  oldNodeAffinity = oldPod.Spec.Affinity.NodeAffinity
}

var mungedNodeAffinity *core.NodeAffinity
if mungedPod.Spec.Affinity != nil {
  mungedNodeAffinity = mungedPod.Spec.Affinity.NodeAffinity
}

if !apiequality.Semantic.DeepEqual(oldNodeAffinity, mungedNodeAffinity) {
  allErrs = append(allErrs, validateNodeAffinityMutation(specPath.Child("affinity").Child("nodeAffinity"), mungedNodeAffinity, oldNodeAffinity)...)
  if mungedPod.Spec.Affinity != nil {
    mungedPod.Spec.Affinity.NodeAffinity = oldNodeAffinity // +k8s:verify-mutation:reason=clone
  }  
}
```